firedrake/pointcloud.py

```diff
@@ -0,0 +1,299 @@
+        # the number of points that will be sent.
+        from_ranks, from_data = build_two_sided(self.mesh.comm, 1, MPI.INT, to_ranks,
+                                                to_data)
+
+        # Create dictionary to hold all receive buffers for point requests from
+        # each process.
+        recv_points_buffers = {}
+        for i in range(0, len(from_ranks)):
+            recv_points_buffers[from_ranks[i]] = np.array(
+                [np.empty(self.mesh.cell_dimension(), dtype=ScalarType)
+                 for _ in range(from_data[i])])
+
+        # Receive all point requests
+
+        self._perform_sparse_communication_round(recv_points_buffers, local_candidates)
+
+        syncPrint("[%d] Point queries requested: %s" % (self.mesh.comm.rank,
+                                                        str(recv_points_buffers)),
+                  comm=self.mesh.comm)
+        syncFlush(comm=self.mesh.comm)
+
+        # Evaluate all point requests and prepare responses
+
+        # Create dictionary to store results.
+        point_responses = {}
+
+        # Evaluate results.
+        for rank, points_buffers in recv_points_buffers.items():
+            point_responses[rank] = self.mesh.locate_cells(np.array([points_buffers]))
+
+        syncPrint("[%d] Point responses: %s" % (self.mesh.comm.rank,
+                                                str(point_responses)),
+                  comm=self.mesh.comm)
+        syncFlush(comm=self.mesh.comm)
+
+        # Receive all responses
+
+        # Create dictionary to hold all output buffers indexed by rank.
+        recv_results = {}
+        # Initialise these.
+        for rank, points in local_candidates.items():
+            # Create receive buffer(s).
+            recv_results[rank] = np.array([np.empty(1, dtype=IntType)
+                                           for _ in range(len(points))])
+
+        self._perform_sparse_communication_round(recv_results, point_responses)
+
+        syncPrint("[%d] Point location request results: %s" % (self.mesh.comm.rank,
+                                                               str(recv_results)),
+                  comm=self.mesh.comm)
+        syncFlush(comm=self.mesh.comm)
+
+        # Process and return results.
+
+        # Iterate through all points. If they have not been located locally,
+        # iterate through each point request reponse to find the element.
+        # Sometimes an element can be reported as found by more than one
+        # process -- in this case, choose the process with the lower rank.
+        for rank, requested_points in local_candidates.items():
+            for j, point in enumerate(requested_points):
+                i = local_candidate_indices[rank][j]
+                curr_rank = located_elements[i][0]
+                result = recv_results[rank][j][0]
+                if (result != -1 and (curr_rank == -1 or rank < curr_rank)):
```

Comment:
I guess we could arbitrarily decide to choose the max rank, rather than the min rank, then we just write:
```
located_elements[i] = (max(rank, curr_rank), result)
```
With no need for the `if`.

Reply:
I think we never used the `point` here, so perhaps:
```python
for rank, result in recv_results.items():
   indices = local_candidate_indices[rank]
   found, = np.where(result != -1)
   for idx in found:
      i = indices[idx]
      loc_rank = located_elements[i, 0]
      if loc_rank == -1 or rank < loc_rank:
          located_elements[i, :] = (rank, result[idx])
```

Reply:
I have went with your second suggestion here -- with the first, if two processes report locating the same point on different elements, would `located_elements` not then be updated with the new result regardless of which rank is higher or am I misunderstanding? Cheers.